### ISSUE TITLE
Mismatched type returned from log_file

### ISSUE THREAD

**Issue by ltfschoen** [daboross/fern#26]
I'm using `fern = "0.5"` in Cargo.toml

When I use the sample code in CLion IDE, the line `.chain(fern::log_file("output.log")?)` displays warning `mismatched types expected Result<File, Error>, found File`

When I go modify the Fern library file **fern-0.5.5/src/lib.rs** and change the signature from `pub fn log_file<P: AsRef<Path>>(path: P) -> io::Result<File> {` to `pub fn log_file<P: AsRef<Path>>(path: P) -> File {` then the warning disappears, and the implementation still outputs to the log file and terminal successfully.

**Comment by daboross** [comment]
I don't believe it should compile if `log_file` is changed to return a `File` directly. Opening a file is an operation that can fail.

Could you elaborate about the mismatched type warnings? This seems like it's possibly an error in the IDE that we could investigate. Provided your function returns a `Result`, `.chain(fern::log_file("output.log")?)` should definitely be correct.

Fern's code does currently compile with no warnings from the rust compiler, and the example which uses `fern::log_file` also compiles without warnings.

**Comment by ltfschoen** [comment]
I'm using CLion 2018.1.1 with macOS Sierra 10.12.6
I have Rust language support installed Version: 0.2.0.2092-181
I have installed the IntelliJ-Rust plugin into the CLion IDE https://intellij-rust.github.io/docs/quick-start.html
The Rust type checker is giving me error E0308 https://doc.rust-lang.org/error-index.html#E0308
The code compiles and runs correctly when I run `cargo run` even if don't change the return value from `io::Result<File>` to `File`, but if I make that change the warning in the IDE goes away

Below is a screenshot of what appears in the CLion IDE:

![screen shot 2018-04-16 at 3 07 07 pm](https://user-images.githubusercontent.com/6226175/38791001-575da466-4188-11e8-9a6e-b44b88cd15fe.png)

Below is screenshot of the lib.rs file with the log_file function that is called by my code:

![screen shot 2018-04-16 at 3 17 47 pm](https://user-images.githubusercontent.com/6226175/38791113-161d479e-4189-11e8-97d9-b2e2f0b11a77.png)

After writing all the above, I decided to set a breakpoint on the line of code that calls log_file, with it changed to the `File` return type to see what that would do, and it displayed the below error:

![screen shot 2018-04-16 at 3 21 28 pm](https://user-images.githubusercontent.com/6226175/38791203-be0abfc2-4189-11e8-8977-0ef87217c846.png)

When I changed it back to `io::Result<File>` return type and ran with breakpoint set it did not result in error.

When I changed it back to `File` once again, built the code with `cargo build` and ran the code with the debugger the same way again (i.e. `cargo run`), strangely it didn't encounter the same error again, even on subsequent attempts. So I think it may be a bug with the IDE...

**Comment by daboross** [comment]
Hmm, it does seem to be some sort of ide error. The compiling output you show in your second screenshot is exactly what I'd expect if you change that code the way your PR did. It needs to return a `Result`.

Have you tried clearing all of CLion's caches and restarting? I'm 90% sure there should be an "invalidate caches & restart" option somewhere which could do this. If that doesn't fix it, there's probably some underlying bug in the rust plugin that needs to be reported at https://github.com/intellij-rust/intellij-rust?

**Comment by ltfschoen** [comment]
I just tried "File > Invalidate Caches / Restart" from the menu but it still shows the same warning. I tried changing lib.rs to `File` again to remove the warning, but it still compiles and runs even in debugging mode without showing any errors in the IDE, which is strange. FYI, this is the code where I'm using it in the 'notary' branch, and i'm running the code with `cargo run -- -mode b`. https://github.com/Drops-of-Diamond/diamond_drops/pull/32.
I'll create an Issue at  https://github.com/intellij-rust/intellij-rust. 
It appears that there are other similar issues previously raised such as this one: https://github.com/intellij-rust/intellij-rust/issues/2461
